### DISCUSSION TITLE
Where possible, intermediate files should be removed

### DISCUSSION THREAD

**Discussion by graeme-winter** [xia2/xia2#591] [Ideas]
A very modest data set still takes up ~ 1GB when processed as we have many copies of "the same" reflection data - in each case just have more columns usually so the later ones are a suoerset of the earlier ones:

```
Grey-Area xds :) $ du -hs $(find . -name '*refl')
 50M	./DEFAULT/scale/16_scaled.refl
 55M	./DEFAULT/scale/14_symmetrized.refl
 50M	./DEFAULT/scale/19_symmetrized.refl
 50M	./DEFAULT/scale/18_scaled.refl
 55M	./DEFAULT/NATIVE/SWEEP1/integrate/12_integrated.refl
 55M	./DEFAULT/NATIVE/SWEEP1/integrate/15_reindexed.refl
788K	./DEFAULT/NATIVE/SWEEP1/integrate/reference_profiles.refl
1.1M	./DEFAULT/NATIVE/SWEEP1/index/4_spot_xds.refl
 44M	./DEFAULT/NATIVE/SWEEP1/index/8_reindexed.refl
 44M	./DEFAULT/NATIVE/SWEEP1/index/5_indexed.refl
 40M	./DEFAULT/NATIVE/SWEEP1/index/2_SWEEP1_strong.refl
 44M	./DEFAULT/NATIVE/SWEEP1/refine/9_refined.refl
 44M	./DEFAULT/NATIVE/SWEEP1/refine/10_refined.refl
 50M	./DataFiles/AUTOMATIC_DEFAULT_scaled.refl
 55M	./DataFiles/AUTOMATIC_DEFAULT_NATIVE_SWEEP1.refl
```

Should we delete redundant data e.g. integrated data before scaling, anything before indexing etc. since in principle the later data are a superset of the earlier? May help users with their disk space.

**Comment by dagewa** [comment]
There was a conversation somewhere, that I now can't find, that might be relevant. The idea proposed then was for xia2 to publish a set of milestone steps during processing so that the interested user could look at, say the spot-finding results, the indexing, the refinement etc. and potentially branch off processing from one of these steps.

This does not need all of the intermediate files, just some of them. In fact, setting up various launch points from a xia2 run would be even easier if lots of the redundant files were removed and only milestone results were kept.

**Reply by graeme-winter** [reply]
My thinking was this:
- refinement output is a superset indexing which is in turn superset of spot finding
- scaling is likewise a superset of symmetry ... etc. to integration

Integration and spot finding are "special" as they create reflections

Milestones / success I like - in event of failure you do not delete anything though, maybe, to aid debugging. 

What is "launch point?" - it sounds relevent here

**Reply by dagewa** [reply]
By "launch point" I mean any state of the processing that xia2 keeps around and "publishes" (in some way) as a state that might be interesting for inspection by the user. CCP4i2 has DIALS image viewer and reciprocal lattice viewer tasks that start from various points in a completed xia2 run:
![xia2](https://user-images.githubusercontent.com/11502083/117266178-e1902580-ae4c-11eb-9f45-b7ae98e9cb61.png)

The mechanism for figuring these points out is a bit shonky and fragile though. It would be easier if these points were better defined, which could be by publication (`milestones.json` or whatever), by tidying up all the superfluous output, or both. It might also help another project... https://github.com/ccp4/DUI/issues/140.

I see your point about failed jobs and keeping everything. Indeed perhaps it is the failed jobs where the milestone publication would be most useful, so the user can quickly see that things were okay at indexing, say, but the global refinement went badly.

I completely agree that tidying up is a good thing as xia2 processing directories can be rather portly (and will only get more so with mmCIF reflections, see #534). Some questions that come to mind:

- How aggressive should the tidying be? Strictly based on contained sets as suggested (might this be only `refined.refl` and `scaled.refl`?) or more based on some idea about natural milestones?
- What about `.expt` files? Should some of these also be removed, or keep all because they are generally small? I can see how it could be useful to start from a different Bravais setting without re-running spot finding, for example.
- Should xia2 publish something like `milestones.json` or let the user or downstream software figure it out?

**Comment by rjgildea** [comment]
Don't we already kind of do this with `record_data_file()` and `record_more_data_file()`, which control which "useful" output files end up in the `DataFiles` directory? I.e. at Diamond we run xia2 in a temporary directory, and only copy the top level `DataFiles`, `LogFiles` etc directories to the final results location.

https://github.com/xia2/xia2/blob/6d6d73003234b735d41866c253c49912472efc04/src/xia2/Handlers/Files.py#L90-L100

**Reply by dagewa** [reply]
I think `DataFiles/` is missing a record of the strong spots prior to integration. It is probably possible to extract this from the integration `.refl` (unless we strip out e.g. unindexed - I can't remember), but it makes viewing spot-finding / indexing results difficult at the moment, if only the files in `DataFiles/` are kept.

**Reply by dagewa** [reply]
ah no, it isn't possible to go back anyway. After integration the `bbox` changes and the `shoebox` is removed.
